[SR-7497] [stdlib] Fix KeyPaths hashing

dmcyk opened the pull request:
- The cause of hashing regression was a shadowed `hasher` argument in component hashing, so the component type/offset wasn't taken into account
- Currently only path components are hashed, so paths with distinct types but properties of equal kind and offset result in same hash values. That differs from equatable implementation as type of two key paths must be the same to consider them equal.
So I thought it'd make sense to include the type metadata pointer while hashing. As far as I understand it, the metadata pointer doesn't change during program execution. 

Regarding code:
 I'm not sure if `type(of: self) as Any.Type` cast is necessary. 
Quickly running REPL trying to do `unsafeBitCast(type(of: Int(0)), to: Int.self)`crashes, casting the `type(of:)` result to `Any.Type` helps though. 
Is this caused by value vs ref type difference? Doing so with a class object works. 
I was looking at `_getTypeName` and other core functions, but I couldn't really find a place where `Any.Type` is guaranteed to be `const Metadata*` and how `Any.Type` differs from e.g. `Int.Type`. 
Is it somehow implemented at SIL level? Would be great if someone could maybe give me a tip where to look for information/code regarding this. 

Resolves [SR-7497](https://bugs.swift.org/browse/SR-7497).

P.S. That's my first attempt at contribution to Swift. Hello everyone 👋

dmcyk committed:
[stdlib] Fix key path components hashing.

Removes shadowed `hasher` argument in component hashing.

dmcyk committed:
[stdlib] Cover type difference in key path hash.

Include type metadata pointer in hashing.

huonw commented:
> Quickly running REPL trying to do unsafeBitCast(type(of: Int(0)), to: Int.self)crashes, casting the type(of:) result to Any.Type helps though.
> Is this caused by value vs ref type difference? Doing so with a class object works.

The metatype object for something that can't be subclassed (or otherwise store interesting information about the type) is zero-sized. The (meta)type `Int.Type` has just one value `Int.self`, meaning it is like `()`/`Void` and has the same size. In contrast, given

```swift
class Super {}
class Sub1: Super {}
class Sub2: Super {}
```

The type `Super.Type` has (at least) three possible values `Super.self`, `Sub1.self` and `Sub2.self`. This is, I believe, represented by the values of `Super.Type` being a pointer to the actual metadata. I believe metatypes of protocol existentials and `Any.Type` are similar, and they are a pointer to the actual information about the type in the existential as you suspect. Someone like @jckarter will have to confirm this, though.

(Hello and welcome!)

dmcyk commented:
ah, I see. I misunderstood it. I thought `Int.Type` was representing the type metadata itself, rather than a meta type. 
I found `init_existential_metatype` instruction in SIL reference, that  says that existential meta types store the type metadata. There's also a note on Metatype types, saying indeed that concrete meta types require no storage. Thanks!

jckarter commented on stdlib/public/core/KeyPath.swift:
```diff
@@ -59,6 +59,7 @@ public class AnyKeyPath: Hashable, _AppendKeyPath {
   ///   of this instance.
   @_effects(releasenone)
   final public func hash(into hasher: inout Hasher) {
+    hasher.combine(unsafeBitCast(type(of: self) as Any.Type, to: Int.self))
```
This could be written `ObjectIdentifier(type(of: self)).hash(into: &hasher)`

dmcyk replied:
Good to know, thanks 👌

jckarter commented:
Nice catch!

dmcyk committed:
[stdlib] Key path safe type hashing.

Uses ObjectIdentifier to extract metatype pointer,
instead of unsafe pointer casting.

lorentey reviewed (approved):
Thanks for spotting this!

lorentey commented:
@swift-ci please smoke test and merge

lorentey commented:
@swift-ci please smoke test and merge

dmcyk commented:
@lorentey my pleasure 😁 I guess this could be merged into 4.2 too, right?

lorentey commented:
@dmcyk Yes, we should get it into 4.2! I'll cherry pick it after it lands on master.
